reactive/kotlinx-coroutines-reactive/src/Await.kt

```diff
@@ -134,29 +135,47 @@ private suspend fun <T> Publisher<T>.awaitOne(
     mode: Mode,
     default: T? = null
 ): T = suspendCancellableCoroutine { cont ->
+    /* This implementation must obey
+    https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/README.md#2-subscriber-code
+    The numbers of rules are taken from there. */
     injectCoroutineContext(cont.context).subscribe(object : Subscriber<T> {
-        private lateinit var subscription: Subscription
+        // It is unclear whether 2.13 implies (T: Any), but if so, it seems that we don't break anything by not adhering
+        private var subscription: Subscription? = null
         private var value: T? = null
         private var seenValue = false
+        private var inTerminalState = false
 
         override fun onSubscribe(sub: Subscription) {
+            /** cancelling the existing subscription due to rule 2.5, though the publisher would either have to
+             * subscribe more than once, which would break 2.12, or leak this [Subscriber]. */
+            subscription?.let {
+                value = null
+                seenValue = false
+                inTerminalState = false
+                it.cancel()
+            }
             subscription = sub
             cont.invokeOnCancellation { sub.cancel() }
-            sub.request(if (mode == Mode.FIRST) 1 else Long.MAX_VALUE)
+            sub.request(if (mode == Mode.FIRST || mode == Mode.FIRST_OR_DEFAULT) 1 else Long.MAX_VALUE)
         }
 
         override fun onNext(t: T) {
+            val sub = subscription.checkInitialized("onNext")
+            subscription.checkInitialized("onNext")
+            if (inTerminalState)
+                gotSignalInTerminalStateException("onNext")
             when (mode) {
                 Mode.FIRST, Mode.FIRST_OR_DEFAULT -> {
-                    if (!seenValue) {
-                        seenValue = true
-                        subscription.cancel()
-                        cont.resume(t)
-                    }
+                    if (seenValue)
+                        // TODO: decide if we want to be lenient here: after all, nothing breaks if this isn't true
+                        moreThanOneValueProvidedException(mode)
```

Comment:
When writing a comment along with a statement, please wrap it in curly braces, otherwise it's a bit hard to read

Reply:
Throwing here breaks `#2.13`. This is expected, but still worth properly documenting to know why it is here even tho it violates the spec.
Also, removing this condition does not trigger any test failures, because double-resume also throws `IllegalStateException`. It's worth verifying the exception message at least once

Reply:
I am now verifying the exception messages. I don't like what I did in order to achieve it, but I like copy-pasting exception messages into tests even less, given how brittle that approach is. Maybe you have some better suggestions?

Reply:
The approach we're consistently using is the following:

- Do not expose exception messages to test in any way (e.g. via `internal` function)
- Instead of strictly comparing the error message as a whole, typically it's enough to do something like `e.message!!.contains("some_key_word_or_meaningful_part_of_the_message")`. 
 
 It simplifies the maintenance a bit, reduces the number of functions/string consts, but it's still brittle. Yet exception messages are rarely changed and the purpose of the test is rather fix the current behaviour